components/tabs/src/sync/record.rs

```diff
@@ -27,7 +29,379 @@ pub struct TabsRecord {
 impl TabsRecord {
     #[inline]
     pub fn from_payload(payload: sync15::Payload) -> Result<Self> {
-        let record: TabsRecord = payload.into_record()?;
-        Ok(record)
+        // Note: We are hand-parsing the tabs payload so that we can provide support for
+        // older clients that may have a stringified float or integer `last_used` value.
+        let client_name: String =
+            parse_string_from_json(&"clientName", payload.data.get("clientName"))?;
+        let tabs: Result<Vec<TabsRecordTab>> = payload
+            .data
+            .get("tabs")
+            .ok_or_else(|| serde_json::Error::custom("missing field `tabs`"))?
+            .as_array()
+            .ok_or_else(|| serde_json::Error::custom("invalid `tabs`, expected sequence"))?
+            .iter()
+            .map(|x| -> Result<TabsRecordTab> {
+                let tabs_obj: Map<String, JsonValue> = x.as_object().unwrap_or(&Map::new()).clone();
+                let title: String = parse_string_from_json(&"title", tabs_obj.get("title"))?;
+                let url_history: Vec<String> = tabs_obj
+                    .get("urlHistory")
+                    .ok_or_else(|| serde_json::Error::custom("missing field `urlHistory`"))?
+                    .as_array()
+                    .ok_or_else(|| {
+                        serde_json::Error::custom("invalid `urlHistory`, expected sequence")
+                    })?
+                    .iter()
+                    .map(|u| -> Result<String> {
+                        Ok(u.as_str()
+                            .ok_or_else(|| {
+                                serde_json::Error::custom(
+                                    "invalid `urlHistory` value, expected string",
+                                )
+                            })?
+                            .to_string())
+                    })
+                    .into_iter()
+                    .collect::<Result<_>>()?;
+                let icon: Result<Option<String>, serde_json::Error> =
+                    if let Some(icon) = tabs_obj.get("icon") {
+                        if let Some(i) = icon.as_str() {
+                            Ok(Some(i.to_string()))
+                        } else {
+                            Err(serde_json::Error::custom("invalid `icon`, expected string"))
+                        }
+                    } else {
+                        Ok(None)
+                    };
+
+                let last_used = parse_last_used(
+                    tabs_obj
+                        .get("lastUsed")
+                        .ok_or_else(|| serde_json::Error::custom("missing field `lastUsed`"))?,
+                )?;
+
+                Ok(TabsRecordTab {
+                    title,
+                    url_history,
+                    icon: icon?,
+                    last_used,
+                })
+            })
+            .into_iter()
+            .collect::<Result<_>>();
```

Comment:
Stylistically, I think it would be cleaner to apply the question-mark here like `.collect::<Result<_>>()?;` rather than later on when you're constructing the `TabsRecord`, then you don't have to repeat the `Result` type when you're defining the `tabs` variable.